File: bindings/ruby/test/test_hyperparameter.rb
[ '../lib', 'lib' ].each { |d| $:.unshift(d) if File::directory?(d) }
require 'minitest/autorun'
require 'cconfigspace'

class CConfigSpaceTestHyperparameter < Minitest::Test
  def setup
    CCS.init
  end

  def test_ordinal_compare
    values = ["foo", 2, 3.0]
    h = CCS::OrdinalHyperparameter::new(values: values)
    assert_equal( 0, h.compare("foo", "foo") )
    assert_equal( -1, h.compare("foo", 2) )
    assert_equal( -1, h.compare("foo", 3.0) )
    assert_equal( 1, h.compare(2, "foo") )
    assert_equal( 0, h.compare(2, 2) )
    assert_equal( -1, h.compare(2, 3.0) )
    assert_equal( 1, h.compare(3.0, "foo") )
    assert_equal( 1, h.compare(3.0, 2) )
    assert_equal( 0, h.compare(3.0, 3.0) )
    assert_raises(CCS::CCSError, :CCS_INVALID_VALUE) { h.compare(4.0, "foo") }
  end

  def test_from_handle_ordinal
    values = ["foo", 2, 3.0]
    h = CCS::OrdinalHyperparameter::new(values: values)
    h2 = CCS::Object::from_handle(h)
    assert_equal( h.class, h2.class )
  end

  def test_ordinal
    values = ["foo", 2, 3.0]
    h = CCS::OrdinalHyperparameter::new(values: values)
    assert_equal( :CCS_HYPERPARAMETER, h.object_type )
    assert_equal( :CCS_ORDINAL, h.type )
    assert_match( /param/, h.name )
    assert( h.user_data.null? )
    assert_equal( "foo", h.default_value )
    assert_equal( :CCS_UNIFORM, h.default_distribution.type )
    assert_equal( values, h.values )
    assert( h.check_value("foo") )
    assert( h.check_value(2) )
    assert( h.check_value(3.0) )
    refute( h.check_value(1.5) )
    v = h.sample
    assert( values.include? v )
    vals = h.samples(100)
    vals.each { |v|
      assert( values.include? v )
    }
  end

  def test_from_handle_categorical
    values = ["foo", 2, 3.0]
    h = CCS::CategoricalHyperparameter::new(values: values)
    h2 = CCS::Object::from_handle(h)
    assert_equal( h.class, h2.class )
  end

  def test_categorical
    values = ["foo", 2, 3.0]
    h = CCS::CategoricalHyperparameter::new(values: values)
    assert_equal( :CCS_HYPERPARAMETER, h.object_type )
    assert_equal( :CCS_CATEGORICAL, h.type )
    assert_match( /param/, h.name )
    assert( h.user_data.null? )
    assert_equal( "foo", h.default_value )
    assert_equal( :CCS_UNIFORM, h.default_distribution.type )
    assert_equal( values, h.values )
    assert( h.check_value("foo") )
    assert( h.check_value(2) )
    assert( h.check_value(3.0) )
    refute( h.check_value(1.5) )
    v = h.sample
    assert( values.include? v )
    vals = h.samples(100)
    vals.each { |v|
      assert( values.include? v )
    }
  end

  def test_from_handle_numerical
    h = CCS::NumericalHyperparameter::new
    h2 = CCS::Object::from_handle(h)
    assert_equal( h.class, h2.class )
  end

  def test_create_numerical
    h = CCS::NumericalHyperparameter::new
    assert_equal( :CCS_HYPERPARAMETER, h.object_type )
    assert_equal( :CCS_NUMERICAL, h.type )
    assert_match( /param/, h.name )
    assert( h.user_data.null? )
    assert_equal( 0.0, h.default_value )
    assert_equal( :CCS_UNIFORM, h.default_distribution.type )
    assert( h.check_value(0.5) )
    refute( h.check_value(1.5) )
    v = h.sample
    assert( v.kind_of?(Float) )
    assert( v >= 0.0 && v < 1.0 )
    vals = h.samples(100)
    vals.each { |v|
      assert( v.kind_of?(Float) )
      assert( v >= 0.0 && v < 1.0 )
    }
  end

  def test_create_numerical_float
    h = CCS::NumericalHyperparameter::float(lower: 0.0, upper: 1.0)
    assert_equal( :CCS_HYPERPARAMETER, h.object_type )
    assert_equal( :CCS_NUMERICAL, h.type )
    assert_match( /param/, h.name )
    assert( h.user_data.null? )
    assert_equal( 0.0, h.default_value )
    assert_equal( :CCS_UNIFORM, h.default_distribution.type )
    assert_equal( :CCS_NUM_FLOAT, h.data_type )
    assert_equal( 0.0, h.lower )
    assert_equal( 1.0, h.upper )
    assert_equal( 0.0, h.quantization )
    assert( h.check_value(0.5) )
    refute( h.check_value(1.5) )
    v = h.sample
    assert( v.kind_of?(Float) )
    assert( v >= 0.0 && v < 1.0 )
    vals = h.samples(100)
    vals.each { |v|
      assert( v.kind_of?(Float) )
      assert( v >= 0.0 && v < 1.0 )
    }
  end

  def test_create_numerical_int
    h = CCS::NumericalHyperparameter::int(lower: 0, upper: 100)
    assert_equal( :CCS_NUMERICAL, h.type )
    assert_match( /param/, h.name )
    assert( h.user_data.null? )
    assert_equal( 0, h.default_value )
    assert_equal( :CCS_UNIFORM, h.default_distribution.type )
    assert_equal( :CCS_NUM_INTEGER, h.data_type )
    assert_equal( 0, h.lower )
    assert_equal( 100, h.upper )
    assert_equal( 0, h.quantization )
    assert( h.check_value(50) )
    refute( h.check_value(150) )
    v = h.sample
    assert( v.kind_of?(Integer) )
    assert( v >= 0 && v < 100 )
    vals = h.samples(100)
    vals.each { |v|
      assert( v.kind_of?(Integer) )
      assert( v >= 0 && v < 100 )
    }
  end

end

File: bindings/python/test/test_hyperparameter.py
import unittest
import sys
sys.path.insert(1, '.')
sys.path.insert(1, '..')
import cconfigspace as ccs

class TestHyperparameter(unittest.TestCase):

  def test_from_handle_discrete(self):
    values = [0, 1.5, 2, 7.2]
    h = ccs.DiscreteHyperparameter(values = values)
    h2 = ccs.Object.from_handle(h.handle)
    self.assertEqual( h.__class__, h2.__class__ )
    self.assertEqual( h.handle, h2.handle )

  def test_discrete(self):
    values = [0.2, 1.5, 2, 7.2]
    h = ccs.DiscreteHyperparameter(values = values)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_DISCRETE, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( 0.2, h.default_value )
    self.assertFalse( h.check_value("foo") )
    v = h.sample()
    self.assertTrue( v in values )
    vals = h.samples(100)
    for v in vals:
      self.assertTrue( v in values )

  def test_serialize_discrete(self):
    values = [0.2, 1.5, 2, 7.2]
    href = ccs.DiscreteHyperparameter(values = values)
    buff = href.serialize()
    h = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_DISCRETE, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( 0.2, h.default_value )
    self.assertFalse( h.check_value("foo") )
    v = h.sample()
    self.assertTrue( v in values )
    vals = h.samples(100)
    for v in vals:
      self.assertTrue( v in values )

  def test_ordinal_compare(self):
    values = ["foo", 2, 3.0]
    h = ccs.OrdinalHyperparameter(values = values)
    self.assertEqual( 0, h.compare("foo", "foo") )
    self.assertEqual( -1, h.compare("foo", 2) )
    self.assertEqual( -1, h.compare("foo", 3.0) )
    self.assertEqual( 1, h.compare(2, "foo") )
    self.assertEqual( 0, h.compare(2, 2) )
    self.assertEqual( -1, h.compare(2, 3.0) )
    self.assertEqual( 1, h.compare(3.0, "foo") )
    self.assertEqual( 1, h.compare(3.0, 2) )
    self.assertEqual( 0, h.compare(3.0, 3.0) )
    self.assertRaises( ccs.Error, h.compare, 4.0, "foo" )

  def test_from_handle_ordinal(self):
    values = ["foo", 2, 3.0]
    h = ccs.OrdinalHyperparameter(values = values)
    h2 = ccs.Object.from_handle(h.handle)
    self.assertEqual( h.__class__, h2.__class__ )
    self.assertEqual( h.handle, h2.handle )

  def test_ordinal(self):
    values = ["foo", 2, 3.0]
    h = ccs.OrdinalHyperparameter(values = values)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_ORDINAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( "foo", h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( values, h.values )
    for v in values:
      self.assertTrue( h.check_value(v) )
    self.assertFalse( h.check_value("bar") )
    v = h.sample()
    self.assertTrue( v in values )
    vals = h.samples(100)
    for v in vals:
      self.assertTrue( v in values )

  def test_serialize_ordinal(self):
    values = ["foo", 2, 3.0]
    href = ccs.OrdinalHyperparameter(values = values)
    buff = href.serialize()
    h = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_ORDINAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( "foo", h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( values, h.values )
    for v in values:
      self.assertTrue( h.check_value(v) )
    self.assertFalse( h.check_value("bar") )
    v = h.sample()
    self.assertTrue( v in values )
    vals = h.samples(100)
    for v in vals:
      self.assertTrue( v in values )

  def test_from_handle_categorical(self):
    values = ["foo", 2, 3.0]
    h = ccs.CategoricalHyperparameter(values = values)
    h2 = ccs.Object.from_handle(h.handle)
    self.assertEqual( h.__class__, h2.__class__ )
    self.assertEqual( h.handle, h2.handle )

  def test_categorical(self):
    values = ["foo", 2, 3.0]
    h = ccs.CategoricalHyperparameter(values = values)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_CATEGORICAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( "foo", h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( values, h.values )
    for v in values:
      self.assertTrue( h.check_value(v) )
    self.assertFalse( h.check_value("bar") )
    v = h.sample()
    self.assertTrue( v in values )
    vals = h.samples(100)
    for v in vals:
      self.assertTrue( v in values )

  def test_serialize_categorical(self):
    values = ["foo", 2, 3.0]
    href = ccs.CategoricalHyperparameter(values = values)
    buff = href.serialize()
    h = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_CATEGORICAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( "foo", h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( values, h.values )
    for v in values:
      self.assertTrue( h.check_value(v) )
    self.assertFalse( h.check_value("bar") )
    v = h.sample()
    self.assertTrue( v in values )
    vals = h.samples(100)
    for v in vals:
      self.assertTrue( v in values )

  def test_from_handle_numerical(self):
    h = ccs.NumericalHyperparameter()
    h2 = ccs.Object.from_handle(h.handle)
    self.assertEqual( h.__class__, h2.__class__ )
    self.assertEqual( h.handle, h2.handle )

  def test_numerical(self):
    h = ccs.NumericalHyperparameter()
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_NUMERICAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( 0.0, h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( ccs.NUM_FLOAT, h.data_type )
    self.assertEqual( 0.0, h.lower )
    self.assertEqual( 1.0, h.upper )
    self.assertEqual( 0.0, h.quantization )
    self.assertTrue( h.check_value(0.5) )
    self.assertFalse( h.check_value(1.5) )
    v = h.sample()
    self.assertIsInstance( v, float )
    self.assertTrue( v >= 0.0 and v < 1.0 )
    vals = h.samples(100)
    for v in vals:
      self.assertIsInstance( v, float )
      self.assertTrue( v >= 0.0 and v < 1.0 )

  def test_serialize_numerical(self):
    href = ccs.NumericalHyperparameter()
    buff = href.serialize()
    h = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_NUMERICAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( 0.0, h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( ccs.NUM_FLOAT, h.data_type )
    self.assertEqual( 0.0, h.lower )
    self.assertEqual( 1.0, h.upper )
    self.assertEqual( 0.0, h.quantization )
    self.assertTrue( h.check_value(0.5) )
    self.assertFalse( h.check_value(1.5) )
    v = h.sample()
    self.assertIsInstance( v, float )
    self.assertTrue( v >= 0.0 and v < 1.0 )
    vals = h.samples(100)
    for v in vals:
      self.assertIsInstance( v, float )
      self.assertTrue( v >= 0.0 and v < 1.0 )

  def test_numerical_float(self):
    h = ccs.NumericalHyperparameter.float(lower = 0.0, upper = 1.0)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_NUMERICAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( 0.0, h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( ccs.NUM_FLOAT, h.data_type )
    self.assertEqual( 0.0, h.lower )
    self.assertEqual( 1.0, h.upper )
    self.assertEqual( 0.0, h.quantization )
    self.assertTrue( h.check_value(0.5) )
    self.assertFalse( h.check_value(1.5) )
    v = h.sample()
    self.assertIsInstance( v, float )
    self.assertTrue( v >= 0.0 and v < 1.0 )
    vals = h.samples(100)
    for v in vals:
      self.assertIsInstance( v, float )
      self.assertTrue( v >= 0.0 and v < 1.0 )

  def test_numerical_int(self):
    h = ccs.NumericalHyperparameter.int(lower = 0, upper = 100)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_NUMERICAL, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    self.assertEqual( 0, h.default_value )
    self.assertEqual( ccs.UNIFORM, h.default_distribution.type )
    self.assertEqual( ccs.NUM_INTEGER, h.data_type )
    self.assertEqual( 0, h.lower )
    self.assertEqual( 100, h.upper )
    self.assertEqual( 0, h.quantization )
    self.assertTrue( h.check_value(50) )
    self.assertFalse( h.check_value(150) )
    v = h.sample()
    self.assertIsInstance( v, int )
    self.assertTrue( v >= 0 and v < 100 )
    vals = h.samples(100)
    for v in vals:
      self.assertIsInstance( v, int )
      self.assertTrue( v >= 0 and v < 100 )

  def test_string(self):
    h = ccs.StringHyperparameter()
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_STRING, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    with self.assertRaises( ccs.Error ):
      h.sample()

  def test_serialize_string(self):
    href = ccs.StringHyperparameter()
    buff = href.serialize()
    h = ccs.Object.deserialize(buffer = buff)
    self.assertEqual( ccs.HYPERPARAMETER, h.object_type )
    self.assertEqual( ccs.HYPERPARAMETER_TYPE_STRING, h.type )
    self.assertTrue( h.name[:5] == "param" )
    self.assertIsNone( h.user_data.value )
    with self.assertRaises( ccs.Error ):
      h.sample()

if __name__ == '__main__':
    unittest.main()
